services/runs.go

```diff
@@ -0,0 +1,315 @@
+	}...)
+
+	if !run.Status.PendingBridge() {
+		return run, fmt.Errorf("Attempting to resume non pending run %s", run.ID)
+	}
+
+	currentTaskRunIndex, ok := run.NextTaskRunIndex()
+	if !ok {
+		return run, fmt.Errorf("Attempting to resume pending run with no remaining tasks %s", run.ID)
+	}
+	currentTaskRun := run.TaskRuns[currentTaskRunIndex]
+
+	var err error
+	if run.Overrides, err = run.Overrides.Merge(input); err != nil {
+		run.TaskRuns[currentTaskRunIndex] = currentTaskRun.ApplyResult(input.WithError(err))
+		*run = run.ApplyResult(input.WithError(err))
+		return run, store.Save(run)
+	}
+
+	currentTaskRun = currentTaskRun.ApplyResult(input)
+	run.TaskRuns[currentTaskRunIndex] = currentTaskRun
+	if currentTaskRun.Status.Finished() && run.TasksRemain() {
+		run.Status = models.RunStatusInProgress
+	} else {
+		*run = run.ApplyResult(input)
+	}
+
+	return run, saveAndTrigger(run, store)
+}
+
+// QueueSleepingTask creates a go routine which will wake up the job runner
+// once the sleep's time has elapsed
+func QueueSleepingTask(
+	run *models.JobRun,
+	store *store.Store,
+) (*models.JobRun, error) {
+	if !run.Status.PendingSleep() {
+		return run, fmt.Errorf("Attempting to resume non sleeping run %s", run.ID)
+	}
+
+	currentTaskRunIndex, ok := run.NextTaskRunIndex()
+	if !ok {
+		return run, fmt.Errorf("Attempting to resume sleeping run with no remaining tasks %s", run.ID)
+	}
+	currentTaskRun := run.TaskRuns[currentTaskRunIndex]
+
+	if !currentTaskRun.Status.PendingSleep() {
+		return run, fmt.Errorf("Attempting to resume sleeping run with non sleeping task %s", run.ID)
+	}
+
+	adapter, err := adapters.For(currentTaskRun.Task, store)
+
+	if err != nil {
+		run.TaskRuns[currentTaskRunIndex] = currentTaskRun.ApplyResult(run.Result.WithError(err))
+		*run = run.ApplyResult(run.Result.WithError(err))
+		return run, store.Save(run)
+	}
+
+	if sleepAdapter, ok := adapter.BaseAdapter.(*adapters.Sleep); ok {
+		return run, performTaskSleep(run, &currentTaskRun, currentTaskRunIndex, sleepAdapter, store)
+	}
+
+	return run, fmt.Errorf("Attempting to resume non sleeping task for run %s (%s)", run.ID, currentTaskRun.Task.Type)
+}
+
+func performTaskSleep(
+	run *models.JobRun,
+	task *models.TaskRun,
+	currentTaskRunIndex int,
+	adapter *adapters.Sleep,
+	store *store.Store) error {
+
+	duration := adapter.Duration()
+	if duration <= 0 {
+		logger.Debugw("Sleep duration has already elapsed, completing task", run.ForLogger()...)
+		task.Status = models.RunStatusCompleted
+		run.TaskRuns[currentTaskRunIndex] = *task
+		run.Status = models.RunStatusInProgress
+		return saveAndTrigger(run, store)
+	}
+
+	// XXX: This is to eliminate data race that occurs because slices share their
+	// underlying array even in copies
+	runCopy := *run
+	runCopy.TaskRuns = make([]models.TaskRun, len(run.TaskRuns))
+	copy(runCopy.TaskRuns, run.TaskRuns)
+
+	go func(run models.JobRun, task models.TaskRun) {
+		logger.Debugw("Task sleeping...", run.ForLogger()...)
```

Comment:
I think this is fine as is, but I did want to point out that this could leak a goroutine in the event that we `Start` or `Stop` the RunChannel, application or JobRunner. In practice, the main goroutine will finish and everything will get cleaned up. Patterns for better clean up are discussed in the Concurrency in Go book and done in HeadTracker.

Reply:
Yeah, for the record I don't think a go routine (especially per task) is the best way to solve this, but this is a continuation of the current implementation. Didn't want to add too much more in one go ;)